### ISSUE TITLE
(Question) Do you imported code from Rust-written hypervisors (VMMs) ?

### ISSUE THREAD

**Issue by hardBSDk** [obhq/obliteration#1094]
It would save a lot of time and import the performance improvements.

The major Rust-written hypervisors:

- crosvm
- Firecracker
- Cloud Hypervisor

**Comment by ultimaweapon** [comment]
AFAIK all of those are a generic VMM or was designed for Linux, which is not suitable for us because:

1. We need to design the kernel to be used with its paravirtualization.
2. Or we need to make the kernel work with the real hardware that it is supported.

For 1 it will be harder for us to optimize the performance or even cause a problem because we need to make it work with its paravirtualization. For 2 the performance will even worst. With our custom made VMM we have a full-control when and where the VM should exit. We also have a full-control how to implement the MMIO that suitable with our workload.

In short, using those VMM likely to cause problems for us. It was designed for the other workload.
